In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
from mlxtend.preprocessing import TransactionEncoder


In [2]:
from google.colab import files
uploaded = files.upload()

Saving Online retail.xlsx to Online retail.xlsx


In [3]:
df = pd.read_excel(r"/content/Online retail.xlsx", header= None)

In [4]:
df.head()

,0
0,"shrimp,almonds,avocado,vegetables mix,green gr..."
1,"burgers,meatballs,eggs"
2,chutney
3,"turkey,avocado"
4,"mineral water,milk,energy bar,whole wheat rice..."


In [5]:
transaction = df[0].dropna().apply(lambda x: x.split(','))

transaction = transaction.tolist()

transaction[:10]


[['shrimp',
  'almonds',
  'avocado',
  'vegetables mix',
  'green grapes',
  'whole weat flour',
  'yams',
  'cottage cheese',
  'energy drink',
  'tomato juice',
  'low fat yogurt',
  'green tea',
  'honey',
  'salad',
  'mineral water',
  'salmon',
  'antioxydant juice',
  'frozen smoothie',
  'spinach',
  'olive oil'],
 ['burgers', 'meatballs', 'eggs'],
 ['chutney'],
 ['turkey', 'avocado'],
 ['mineral water', 'milk', 'energy bar', 'whole wheat rice', 'green tea'],
 ['low fat yogurt'],
 ['whole wheat pasta', 'french fries'],
 ['soup', 'light cream', 'shallot'],
 ['frozen vegetables', 'spaghetti', 'green tea'],
 ['french fries']]

In [6]:
te = TransactionEncoder()
te_ary = te.fit(transaction).transform(transaction)
basket_df = pd.DataFrame(te_ary, columns=te.columns_)
basket_df

,asparagus,almonds,antioxydant juice,asparagus,avocado,babies food,bacon,barbecue sauce,black tea,blueberries,...,turkey,vegetables mix,water spray,white wine,whole weat flour,whole wheat pasta,whole wheat rice,yams,yogurt cake,zucchini
0,False,True,True,False,True,False,False,False,False,False,...,False,True,False,False,True,False,False,True,False,False
1,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
3,False,False,False,False,True,False,False,False,False,False,...,True,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,True,False,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7496,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
7497,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
7498,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False
7499,False,False,False,False,False,False,False,False,False,False,...,False,False,False,False,False,False,False,False,False,False


In [14]:
import warnings
warnings.filterwarnings('ignore', category=DeprecationWarning, module='jupyter_client')

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

In [29]:
from mlxtend.frequent_patterns import apriori

frequent_itemsets = apriori(basket_df, min_support=0.02, use_colnames=True)
print(frequent_itemsets)

      support                           itemsets
0    0.020397                          (almonds)
1    0.033329                          (avocado)
2    0.033729                         (brownies)
3    0.087188                          (burgers)
4    0.030129                           (butter)
..        ...                                ...
98   0.020131  (mineral water, whole wheat rice)
99   0.022930             (olive oil, spaghetti)
100  0.025197              (spaghetti, pancakes)
101  0.021197                (shrimp, spaghetti)
102  0.020931              (tomatoes, spaghetti)

[103 rows x 2 columns]


In [30]:
from mlxtend.frequent_patterns import association_rules

rules = association_rules(frequent_itemsets, metric="lift", min_threshold=1)
display(rules)

,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
0,(burgers),(eggs),0.087188,0.179709,0.028796,0.330275,1.837830,1.0,0.013128,1.224818,0.499424,0.120941,0.183552,0.245256
1,(eggs),(burgers),0.179709,0.087188,0.028796,0.160237,1.837830,1.0,0.013128,1.086988,0.555754,0.120941,0.080026,0.245256
2,(burgers),(french fries),0.087188,0.170911,0.021997,0.252294,1.476173,1.0,0.007096,1.108844,0.353384,0.093168,0.098160,0.190499
3,(french fries),(burgers),0.170911,0.087188,0.021997,0.128705,1.476173,1.0,0.007096,1.047650,0.389069,0.093168,0.045482,0.190499
4,(mineral water),(burgers),0.238368,0.087188,0.024397,0.102349,1.173883,1.0,0.003614,1.016889,0.194486,0.081009,0.016609,0.191083
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
89,(pancakes),(spaghetti),0.095054,0.174110,0.025197,0.265077,1.522468,1.0,0.008647,1.123778,0.379218,0.103279,0.110144,0.204897
90,(shrimp),(spaghetti),0.071457,0.174110,0.021197,0.296642,1.703760,1.0,0.008756,1.174209,0.444850,0.094474,0.148363,0.209194
91,(spaghetti),(shrimp),0.174110,0.071457,0.021197,0.121746,1.703760,1.0,0.008756,1.057260,0.500143,0.094474,0.054159,0.209194
92,(tomatoes),(spaghetti),0.068391,0.174110,0.020931,0.306043,1.757755,1.0,0.009023,1.190117,0.462740,0.094465,0.159746,0.213129


Based on the association rules provided, here is an analysis of the purchasing patterns and customer behavior insights.

### 1. High-Level Summary

The dataset reveals two distinct types of purchasing behaviors: **Meal Preparation** (ingredients bought together to make a specific dish) and **Staple Accumulation** (essential items added to the basket regardless of the main purchase).

* **Most Influential Item:** `Spaghetti` (The anchor for meal-specific rules).
* **Most Common Accompaniment:** `Mineral Water` (The universal add-on).
* **Strongest Association:** `Ground Beef`  `Spaghetti` (Lift: 2.29).

---

### 2. Key Pattern Clusters

#### A. The "Pasta Night" Phenomenon (Strongest Correlation)

The strongest rules in your entire dataset revolve around cooking pasta. The data suggests that when customers buy specific proteins or sauces, they are almost certainly making a pasta dish.

* **The Rules:**
* **Rule 64:** `Ground Beef`  `Spaghetti` (**Lift: 2.29**, Confidence: ~40%)
* **Rule 86:** `Olive Oil`  `Spaghetti` (**Lift: 2.00**, Confidence: ~35%)
* **Rule 92:** `Tomatoes`  `Spaghetti` (**Lift: 1.76**, Confidence: ~31%)
* **Rule 90:** `Shrimp`  `Spaghetti` (**Lift: 1.70**, Confidence: ~30%)


* **Insight:** "Spaghetti" is not just a product; it is a destination category. Customers buying ground beef, olive oil, or tomatoes are statistically very likely to be planning a specific meal (Bolognese or Seafood Pasta).
* **Actionable Advice:** Place parmesan cheese, garlic bread, or basil stations directly next to the Ground Beef and Fresh Tomato sections, not just in the pasta aisle.

#### B. The "Mineral Water" Default

`Mineral Water` appears as a consequent in a massive number of rules (Burgers, Cake, Chicken, Chocolate, Pancakes, etc.).

* **The Rules:**
* **Rule 78:** `Soup`  `Mineral Water` (**Lift: 1.91**)
* **Rule 54:** `Frozen Vegetables`  `Mineral Water` (**Lift: 1.57**)
* **Rule 11:** `Chicken`  `Mineral Water` (**Lift: 1.59**)


* **Insight:** Mineral water behaves as a "basket filler." It has high support but is often bought alongside unrelated items. The high association with **Soup** (Lift 1.91) is particularly interesting—it suggests a specific "Health/Light Meal" shopper profile who values hydration and light food.

#### C. The "Heavy Breakfast" or "Keto" Shopper

There is a surprisingly strong correlation between Burgers and Eggs.

* **The Rule:**
* **Rule 0:** `Burgers`  `Eggs` (**Lift: 1.84**)


* **Insight:** While eggs are typically breakfast and burgers are dinner, their strong association suggests a specific dietary pattern (potentially high-protein/low-carb diets) or a household purchasing distinct protein staples simultaneously.

#### D. The "Creamy Casserole" Pattern

* **The Rule:**
* **Rule 52:** `Frozen Vegetables`  `Milk` (**Lift: 1.91**)


* **Insight:** This is a very specific cooking behavior. The high lift indicates that when people buy frozen veggies, they are frequently buying milk, likely to create sauces, soups, or casseroles.

---

### 3. Metric-Based Interpretation

To understand *why* these rules matter, we look at the three key metrics:

| Rule Pair | Metric Focus | Interpretation |
| --- | --- | --- |
| **Ground Beef  Spaghetti** | **High Lift (2.29)** | **This is a dominant strategy.** The purchase of beef *drastically* increases the probability of buying spaghetti. These items are interdependent. |
| **Mineral Water  Whole Wheat Rice** | **Low Confidence (8%)** | **Weak prediction.** Even though the Lift is positive (1.44), knowing someone bought water only gives you an 8% certainty they will buy rice. This is not a strong rule for promotions. |
| **Burgers  Eggs** | **High Confidence (33%)** | **Reliable habit.** 1 in 3 people who buy burgers also buy eggs. This is a safe bet for bundle pricing. |

### 4. Business Recommendations

1. **Dynamic Bundling:** Create a **"Sunday Dinner Bundle"** offering a discount if Ground Beef, Spaghetti, and Tomato Sauce are bought together. The data proves these naturally go together.
2. **Store Layout:**
* **Cross-Merchandising:** Move a display of `Spaghetti` to the `Meat` department (specifically near Ground Beef).
* **Impulse Buys:** Since `Mineral Water` is bought with *everything* (Cake, Chicken, Pancakes), place single-serve water bottles at the checkout of every aisle, not just the beverage aisle.


3. **Inventory Planning:** If you run a promotion on `Olive Oil`, expect a surge in `Spaghetti` sales. Ensure stock levels for pasta are adjusted up when oil or beef is discounted.